[structured] Preserve computed elements from meta func to impl

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61746

**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

Differential Revision: [D30407831](https://our.internmc.facebook.com/intern/diff/D30407831)

Commit:
[WIP] Preserve computed elements from meta func to impl

[ghstack-poisoned]

Commit:
Update on "[WIP] Preserve computed elements from meta func to impl"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -78,11 +77,13 @@ void check_allany_for_meta(
 
 TORCH_META_FUNC2(all, dim)(const Tensor& self, int64_t dim, bool keepdim) {
-  check_allany_for_meta(*this, "all", self, dim, keepdim);
+  this->dim = maybe_wrap_dim(dim, self.dim());
```
mmmm. Let's enforce the convention that all members have trailing underscores, it will make it more obvious what's going on here

ezyang commented:
Some more philosophical questions that I've wondered about for this task (they may be insoluble, just trying to put some seeds in your brain):

* Is preservation of computed elements in this way an indication that the level of abstraction for the arguments in the function is incorrect? For example, you could imagine rewriting the example function "all" here into two functions, "all" (with the same semantics as before), and "_explicit_all" (where dim is guaranteed to be wrapped correctly). What are the downsides of doing this? (Overhead?) Which one ought to be exposed to external backends? Would this work in all cases?
* Is there a way to do this computed element preservation without forcing them to be spilled to memory (as is the case for class members, due to class ABI requirements)?
* Is there a way to make sure implementations use `this->dim` (in your example) and not the explicitly passed in dim?
* When is it appropriate vs inappropriate to add a field like this?

Commit:
Update on "[WIP] Preserve computed elements from meta func to impl"




[ghstack-poisoned]

Commit:
Update on "[WIP] Preserve computed elements from meta func to impl"




[ghstack-poisoned]

Commit:
Update on "[WIP] Preserve computed elements from meta func to impl"




[ghstack-poisoned]

Commit:
Update on "[WIP] Preserve computed elements from meta func to impl"




[ghstack-poisoned]

Commit:
Update on "[WIP] Preserve computed elements from meta func to impl"




[ghstack-poisoned]

Commit:
Update on "[WIP] Preserve computed elements from meta func to impl"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/TensorMeta.h:
```diff
@@ -26,6 +26,9 @@ namespace impl {
 #define TORCH_META_FUNC2(name, overload) void structured_##name##_##overload::meta
 
+#define TORCH_PRECOMPUTE_META_FUNC(name) structured_##name::precompute_out structured_##name::meta
+#define TORCH_PRECOMPUTE_META_FUNC2(name, overload) structured_##name##_##overload::precompute_out structured_##name##_##overload::meta
```
I wonder if there actually is any reason to change the naming convention here, since, IIUC, precompute meta and meta are mutually exclusive (so it wouldn't be like you would have a naming conflict)

ezyang replied:
reupping

SplitInfinity replied:
Which naming convention did I change? New macros are required because the return type for meta functions of kernels that have precompute are different.

ezyang replied:
oh true. I guess we could unconditionally define `structured_##name::precompute_out` (maybe as `return_type`) and then just typedef it to void when it's not precomputed he he

ezyang replied:
After your last comment, I don't feel too strongly about this, and wouldn't block on this

ezyang commented on aten/src/ATen/native/AveragePool2d.cpp:
```diff
@@ -8,59 +8,88 @@ namespace at {
   }
+
+  return {
+        kH=kH,
```
huh, the dots are optional? Should we prefer if people specify dot or not?

SplitInfinity replied:
No, I think what's happening is that the `kH` declared above is being assigned to itself and then used to make the struct. The dots are definitely required. But on that note, I can't reproduce the error that is supposed to be emitted with this syntax: https://godbolt.org/z/Y6dxh1Eqd

I remember clearly that we did manage to write code that produces that error during a VC, but I can't seem to do it again. I tried gcc and clang both.

ezyang replied:
blegh. Well I guess we'll just have to write a little lint rule for this. Maybe clang tidy can catch it.

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -69,20 +69,33 @@ void check_allany_for_meta(
+   dim = maybe_wrap_dim(dim, self.dim());
+   check_allany_for_meta(*this, "all", self, dim, keepdim);
+   return {
+     .dim_=dim
```
Is there anything stopping us from just calling this dim (no trailing underscore?)

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1250,11 +1263,10 @@ Tensor all(const Tensor& self) {
+  auto iter = get_allany_iter(self, result, dim_, keepdim);
   auto mut_result = const_cast<Tensor&>(result);
-  if (!_dimreduce_return_trivial(mut_result, self, 1, dim, keepdim)) {
+  if (!_dimreduce_return_trivial(mut_result, self, 1, dim_, keepdim)) {
```
Here, there is no more conflict, so might as well just call it dim lol

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -487,6 +487,10 @@
+  precomputed:
+    elements: int64_t dim_
+    replace:
+      - dim -> dim_
```
I guess it's because this would look weird?

SplitInfinity replied:
Yeah, that's why I included the underscore.

ezyang commented on tools/codegen/model.py:
```diff
@@ -1,7 +1,7 @@
 
 from dataclasses import dataclass
-from typing import List, Dict, Optional, Iterator, Tuple, Set, NoReturn, Sequence, Callable, Union
+from typing import Any, List, Dict, Optional, Iterator, Tuple, Set, NoReturn, Sequence, Callable, Union
```
danger danger will robinson! (lol)

ezyang replied:
ok seems justified

ezyang commented on tools/codegen/model.py:
```diff
@@ -229,6 +229,8 @@ class NativeFunction:
     # changes the semantics of set_output to call the parent class.
     structured_inherits: Optional[str]
 
+    precomputed: Optional['Precompute']
```
comments plz

ezyang commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1503,40 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+
+@dataclass(frozen=True)
+class Precompute:
+    elements: List[Argument]
```
s/Argument/PrecomputedElement/ right?

SplitInfinity replied:
Yes! I tried to use `Argument` first to avoid introducing another class, but got stuck because `Argument` parses JIT types, not C++ types.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -487,6 +487,10 @@
   device_check: NoCheck   # TensorIterator
   structured: True
+  precomputed:
+    elements: int64_t dim_
```
I can't remember: did we decide we were going to have elements be JIT schema types or not?

SplitInfinity replied:
No, I think we said we need them to be C++ types because of something CUDA related that i can't remember. I did some Googling and found this post on StackOverflow that states that [operations on 64-bit integers are synthesized using multiple 32-bit operations](https://stackoverflow.com/questions/16845644/64-bit-number-support-in-cuda/16845794), so there could be performance benefits to using 32-bit integers. In the case of `avg_pool2d` for example, the precomputed elements are used to calculate `int` loop bounds, so that explanation kind of makes sense.

However, that post is 8 years old, so it could be out of date.

ezyang replied:
Ah yes, I remember this. But I think I also said something like, it might still be OK to store these as 64-bit and then force the CUDA kernel to truncate them (and it wouldn't be necessary on CPU). Not clear if other kernels would have similar problems or not tho...

SplitInfinity replied:
Okay, so the CPU kernel uses `int64_t` for the actual loop variables derived from these `int` parameters:

https://github.com/pytorch/pytorch/blob/855e8f2b179e3cb7c0e11f9c7757acd610a2a6b4/aten/src/ATen/native/cpu/AvgPoolKernel.cpp#L51-L54

but the CUDA kernel uses `int`:

https://github.com/pytorch/pytorch/blob/855e8f2b179e3cb7c0e11f9c7757acd610a2a6b4/aten/src/ATen/native/cuda/AveragePool2d.cu#L37-L40

so I think your proposal makes sense. Let me try replacing the C++ types in `native_functions.yaml` with JIT types and seeing how it goes.

ezyang commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1503,40 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+    ty: str
+
+    @staticmethod
+    def build(src: str) -> 'PrecomputedElement':
```
BTW, the preexisting nomenclature in the file is "parse" as opposed to "build" (since it's parsing a string)

ezyang commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1503,40 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+@dataclass(frozen=True)
+class PrecomputedElement:
+    name: str
+    ty: str
```
This is one of those places where telling people in a comment that this is a C++ type specifically is really, really helpful semantic information for downstream users. I might go as far to say you should call this `cpp_ty` and not just ty. (This is moot if we decide to do JIT schema types insetad)

ezyang commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1503,40 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+    @staticmethod
+    def build(src: str) -> 'PrecomputedElement':
+        ty, name = src.split(" ")
+        return PrecomputedElement(name=name, ty=ty)
```
In a bunch of the other parsing code we have self validation code that looks like this:

```
        r = Return(
            name=name,
            type=type,
            annotation=annotation,
        )
        assert str(r) == arg, f'{str(r)} != {arg}'
        return r
```

consider adding it here too!

SplitInfinity replied:
Very cool, but how is `__str__` defined on `PrecomputedElement`? Is it added by the `@dataclass` decorator?

ezyang commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1503,40 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+@dataclass(frozen=True)
+class Precompute:
+    elements: List[Argument]
+    replace: Optional[Dict[str, List[Argument]]]
```
It seems simpler to just not make this Optional at all

ezyang replied:
Also not sure why the target is Argument; there are no types in the substitution

SplitInfinity replied:
Ah yes, an empty map does also evaluate to `False`...

Commit:
Update on "[WIP] Preserve computed elements from meta func to impl"




[ghstack-poisoned]

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

[ghstack-poisoned]

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

[ghstack-poisoned]

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/cuda/AveragePool2d.cu:
```diff
@@ -232,39 +232,33 @@ __global__ void avg_pool2d_backward_out_cuda_frame_nhwc(const int nthreads,
+ int dH,
+ int dW,
+ int padH,
+ int padW,
```
you gotta update these to int64_t, and then safe downcast them to int in the bottom

SplitInfinity replied:
Yep, I was going to do this next and didn't request review on purpose for that reason (and to see what else failed on CUDA).

ezyang commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1509,66 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+    # The name of the precomputed element.
+    name: str
+    # The C++ type of the precomputed element.
+    cpp_ty: str
```
since you switched these to JIT type, this should get updated too right?

SplitInfinity replied:
Yep

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8809,6 +8817,12 @@
+    replace:
+      - kernel_size -> kH, kW
+      - stride -> dH, dW
+      - padding -> padH, padW
```
Here's a thought; what if we change this syntax to:

```
precomputed:
      - kernel_size -> int kH, int kW
      - stride -> int dH, int dW
      - padding -> int padH, int padW
```

More compact and there is no information loss

SplitInfinity replied:
After my change to how `Precompute` stores data internally, this makes sense.

ezyang commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1509,66 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+class Precompute:
+    # A map from kernel argument name -> a list of precomputed
+    # elements that replaces/supersedes it.
+    replace: Dict[str, List[Argument]]
```
The syntax change was suggested by your data representation here, which doesn't separately store elements lol

ezyang commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1509,66 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+        # All precomputed elements should be mentioned at least once on the RHS
+        # of an entry in the "replace" field.
+        assert not elements_set
+
```
This is a great place to add the assert that the parsing is reversible ;) ;) ;)

ezyang commented on tools/codegen/dest/native_functions.py:
```diff
@@ -23,13 +23,36 @@ def gen_unstructured(f: NativeFunction, backend_index: BackendIndex) -> Optional
+                    impl_args_replaced.append(f"{cpp_type.cpp_type(strip_ref=True)} {replacement.name}")
+            else:
+                # If not, push a as is.
+                impl_args_replaced.append(a.decl())
```
Now that you've got honest to goodness Argument for these impl arguments, we can shuffle this logic for precompute handling into `structured.impl_arguments`, because that's the right place for it to go :)

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -599,13 +607,30 @@ def generate_defn(cpp_sig: CppSignature) -> str:
-                        method=False
-                    )
-                )
+                impl_exprs_gen = (e.expr for e in translate(context, structured.impl_arguments(self.g), method=False))
```
no change necessary here, but fwiw, I'd probably personally just have this be a list rather than a generator, because the risk of accidentally using the generator multiple times is too big.

ezyang replied:
Also this is predicated on `impl_arguments` actually returning the correct set of required arguments. So if you do the refactoring I suggested you actually have to do it this way, so it's not that big brain anymore

ezyang commented:
This is looking good. Is there any lurking troubles in the CI failures?

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

[ghstack-poisoned]

SplitInfinity commented:
> This is looking good. Is there any lurking troubles in the CI failures?

No, the only errors were that the CUDA impl function signature didn't match (because I didn't declare the precomputed parameters as `int64_t` and some local variable declarations that were missing type specifiers (probably some copy/paste error).

I've addressed all of the comments except for moving the logic that generates the declaration of `impl` into `structured.impl_arguments` and then using `translate` to generate the `impl` call. I'll do those two things tomorrow and then this should be good to go.

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -109,16 +109,18 @@ void check_all_any(const char* name, const Tensor& self, const Tensor& result) {
+TORCH_PRECOMPUTE_META_FUNC2(all, dim)(const Tensor& self, int64_t dim, bool keepdim) {
   check_all_any("all", self, maybe_get_output());
   auto out_dtype = get_result_or_bytebool_dtype(self, maybe_get_output());
   resize_reduction(*this, self, dim, keepdim, out_dtype);
```
This was added while I was working on this PR and calls `maybe_wrap_dim` internally, so we're still calling that twice (in the meta now instead of once in the meta and once in the impl). Not sure what to do here; having `resize_reduction` return the result doesn't make sense.

ezyang replied:
bloop

SplitInfinity commented:
I've addressed all comments. I'm going to wait a few hours for test signal before I request review.

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

[ghstack-poisoned]

SplitInfinity commented on tools/codegen/model.py:
```diff
@@ -1496,3 +1509,38 @@ def parse_returns(return_decl: str) -> Tuple[Return, ...]:
+    replace: Dict[str, List[Argument]]
+
+    @staticmethod
+    def parse(src: List[str]) -> 'Precompute':
```
This causes a mypy error because the annotated type of the object that gets passed to this is `object`. Should I use `Any`/`object` instead and add `isinstance` checks to make sure it is a `list` of `str`?

ezyang replied:
yep. this is because the yaml is completely untyped and you gotta test if it's the right type

ezyang reviewed (approved):
This is really slick, thanks a lot!

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30407831).

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

Differential Revision: [D30407831](https://our.internmc.facebook.com/intern/diff/D30407831)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30407831).

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

Differential Revision: [D30407831](https://our.internmc.facebook.com/intern/diff/D30407831)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30407831).

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

Differential Revision: [D30407831](https://our.internmc.facebook.com/intern/diff/D30407831)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30407831).

ezyang commented on aten/src/ATen/native/AveragePool2d.cpp:
```diff
@@ -82,14 +82,7 @@ TORCH_PRECOMPUTE_META_FUNC(avg_pool2d)
-        .padH=padH,
-        .padW=padW
-  };
+  return {kH, kW, dH, dW, padH, padW};
```
uh oh

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

Differential Revision: [D30407831](https://our.internmc.facebook.com/intern/diff/D30407831)

[ghstack-poisoned]

SplitInfinity commented:
Okay, so I tried the fluent approach with a templated class. It went pretty smoothly for the most part, but I commented on the diff with some questions I have. Here's an example of a `precompute_out` struct generated by this version of the PR:
```
struct TORCH_API structured_avg_pool2d : public at::impl::MetaBase {
  template <bool A = false, bool B = false, bool C = false, bool D = false, bool E = false, bool F = false>
  struct TORCH_API precompute_out {
    precompute_out<true, B, C, D, E, F> set_kH(int64_t value) {
      TORCH_INTERNAL_ASSERT(A == false);
      precompute_out<true, B, C, D, E, F> ret;
      ret.kH = value;
      ret.kW = this->kW;  
      ret.dH = this->dH;  
      ret.dW = this->dW;
      ret.padH = this->padH;
      ret.padW = this->padW;
      return ret;             
    }   
    
    precompute_out<A, true, C, D, E, F> set_kW(int64_t value) {
      TORCH_INTERNAL_ASSERT(B == false);
      precompute_out<A, true, C, D, E, F> ret;
      ret.kH = this->kH;
      ret.kW = value;
      ret.dH = this->dH;  
      ret.dW = this->dW;
      ret.padH = this->padH;
    }

    ...

    int64_t kH;
    int64_t kW;
    int64_t dH;
    int64_t dW;         
    int64_t padH;
    int64_t padW;       
};

I left the other setters out but I think you get the idea.

SplitInfinity commented on tools/codegen/gen.py:
```diff
@@ -456,9 +456,97 @@ def compute_meta_function_declaration(g: NativeFunctionsGroup) -> Optional[str]:
+        meta_return = "void"
+        precomputed = g.out.precomputed if g.structured else None
+
+        if precomputed:
```
The code added below looks really ugly:
```
struct TORCH_API structured_avg_pool2d : public at::impl::MetaBase {
  
                template <bool A = false, bool B = false, bool C = false, bool D
                struct TORCH_API precompute_out {
                    
                    precompute_out<true, B, C, D, E, F> set_kH(int64_t value) {
                        TORCH_INTERNAL_ASSERT(A == false);
                        precompute_out<true, B, C, D, E, F> ret;
ret.kH = value;     
ret.kW = this->kW;      
ret.dH = this->dH;      
ret.dW = this->dW;
ret.padH = this->padH;
ret.padW = this->padW;  
return ret;
                    }
```

Is there a way to indent these blocks properly (other than adding tab characters manually)?

SplitInfinity commented on aten/src/ATen/native/AveragePool2d.cpp:
```diff
@@ -82,7 +82,7 @@ TORCH_PRECOMPUTE_META_FUNC(avg_pool2d)
   }
 
-  return {kH, kW, dH, dW, padH, padW};
+  return TORCH_PRECOMPUTE_STRUCT(avg_pool2d)().set_kH(kH).set_kW(kW).set_dH(dH).set_dW(dW).set_padH(padH).set_padW(padW);
```
So the old `{kH, kW, dH, dW, padH, padW}` syntax still works here because the precompute struct is declared as a `struct`. I would like to disallow that; could do so by making the `struct` a `class` where everything is public?

ezyang replied:
It might be sufficient to define a constructor, not sure

SplitInfinity commented on tools/codegen/gen.py:
```diff
@@ -460,26 +460,92 @@ def compute_meta_function_declaration(g: NativeFunctionsGroup) -> Optional[str]:
+                # element that is set by this method is false on the
+                # class corresponding to the object that `this` points to.
+                # This ensures that each element can be set only once.
+                assert_stmt = f"TORCH_INTERNAL_ASSERT({precomputed_template_parameters[i]} == false);"
```
I tried to use `std::enable_if` to disable the corresponding method altogether when this condition is not met, but that led to not being able to instantiate `precompute_out<true, true, ..., true>` (i.e. the final instance will all members set). Hopefully I did something dumb and there is a way to convert this assert into a a compile time error.

SplitInfinity replied:
Oh, what about `static_assert`? :O

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30407831).

SplitInfinity commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -109,16 +109,18 @@ void check_all_any(const char* name, const Tensor& self, const Tensor& result) {
   check_all_any("any", self, maybe_get_output());
   auto out_dtype = get_result_or_bytebool_dtype(self, maybe_get_output());
   resize_reduction(*this, self, dim, keepdim, out_dtype);
+  return TORCH_PRECOMPUTE_STRUCT2(any, dim)().set_dim(maybe_wrap_dim(dim, self.dim()));
```
I dislike this extra pair of `()` after the macro but I think it would be even worse to include them in the macro itself. The macro is supposed to expand into the correct type name and should not do anything more than that.

SplitInfinity commented on tools/codegen/gen.py:
```diff
@@ -456,9 +456,97 @@ def compute_meta_function_declaration(g: NativeFunctionsGroup) -> Optional[str]:
+                    else:
+                        construction_stmts.append(f"ret.{elem.name} = this->{elem.name};")
+
+                construction_stmts.append("return ret;")
```
No; how would I go about determining that?

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

Differential Revision: [D30407831](https://our.internmc.facebook.com/intern/diff/D30407831)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30407831).

Commit:
Update on "[structured] Preserve computed elements from meta func to impl"


**Summary**
This commit introduces a new feature for structured kernels that allows
kernels to declare quantities as "precomputed" in
`native_functions.yaml`, compute them once in the `meta` function and
reuse them again in the `impl`. The names and types of these quantities
are used to generate code for a struct containing them that the `meta`
function must return. In the case of a handful of surveyed kernels
(`all,`, `any`, `avg_pool2d`), these quantities that are used both in
the `meta` and `impl` have the same meaning as certain kernel arguments
and in fact supersede them. Accordingly, the correspondence between a
kernel argument and the precomputed elements that supersede it is also
captured in `native_functions.yaml`. This information is used to unpack
the struct returned by `meta` and pass its contents correctly to the
`impl` function.

The primary goal is to avoid recompute and enhance developer experience
(e.g. sometimes people can forget to compute these elements while
porting a kernel).

Differential Revision: [D30407831](https://our.internmc.facebook.com/intern/diff/D30407831)

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30407831).
